{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "hw2.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "mFvUKqx0Lss5",
        "DvXwwE7zf0Qw",
        "PIBVoyeyf9UJ",
        "GXcCRnv6Nyum",
        "QSwpykMdN2bu",
        "VdoQIFL0SFH3",
        "-9212_KzgIqw",
        "oHbqE8XKtlyC"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Подготовка"
      ],
      "metadata": {
        "id": "mFvUKqx0Lss5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем наши контиги из предыдущей домашки "
      ],
      "metadata": {
        "id": "K23ThE-1JaQu"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "JYvnz2etG6xv",
        "outputId": "9d9a13d5-bd02-4485-8af1-c10cc99f9222"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File ‘scaffold.fasta’ already there; not retrieving.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget -nc https://raw.githubusercontent.com/cortan122/hse21_hw1/master/data/scaffold.fasta"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем GeneMark и ключ от него (долой proprietary software!! ☭)"
      ],
      "metadata": {
        "id": "U0n7QYGAJkIR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -nc http://topaz.gatech.edu/GeneMark/tmp/GMtool_zAklJ/gm_key_64.gz\n",
        "!wget -nc http://topaz.gatech.edu/GeneMark/tmp/GMtool_zAklJ/gms2_linux_64.tar.gz\n",
        "!gzip -d gm_key_64.gz\n",
        "!tar -xzvf gms2_linux_64.tar.gz\n",
        "!cp -v gm_key_64 ~/.gmhmmp2_key"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SJ1D_NtCHUYz",
        "outputId": "a621a929-82fd-464d-8504-aed7261b7b3b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-13 17:44:49--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_zAklJ/gm_key_64.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 86 [application/x-gzip]\n",
            "Saving to: ‘gm_key_64.gz’\n",
            "\n",
            "\rgm_key_64.gz          0%[                    ]       0  --.-KB/s               \rgm_key_64.gz        100%[===================>]      86  --.-KB/s    in 0s      \n",
            "\n",
            "2021-12-13 17:44:49 (13.3 MB/s) - ‘gm_key_64.gz’ saved [86/86]\n",
            "\n",
            "File ‘gms2_linux_64.tar.gz’ already there; not retrieving.\n",
            "\n",
            "gzip: gm_key_64 already exists; do you wish to overwrite (y or n)? ^C\n",
            "gms2_linux_64/\n",
            "gms2_linux_64/gms2.pl\n",
            "gms2_linux_64/README_GMS2.txt\n",
            "gms2_linux_64/gmhmmp2_key\n",
            "gms2_linux_64/mgm_4.mod\n",
            "gms2_linux_64/gmhmmp2\n",
            "gms2_linux_64/INSTALL\n",
            "gms2_linux_64/mgm_11.mod\n",
            "gms2_linux_64/biogem\n",
            "gms2_linux_64/LICENSE\n",
            "gms2_linux_64/mgm_25.mod\n",
            "gms2_linux_64/compp\n",
            "gms2_linux_64/mgm_15.mod\n",
            "'gm_key_64' -> '/root/.gmhmmp2_key'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем данные бактерии T.oleivorans"
      ],
      "metadata": {
        "id": "Br66yg9TKtUC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh | sh -\n",
        "!~/edirect/efetch -db nuccore -id HF680312 -format gb > T_oleivorans_MIL_1.gbk\n",
        "!~/edirect/efetch -db nuccore -id HF680312 -format gene_fasta > T_oleivorans_MIL_1.genes.fasta\n",
        "!~/edirect/efetch -db nuccore -id HF680312 -format fasta_cds_aa > T_oleivorans_MIL_1.proteins.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wYShCViVKwky",
        "outputId": "89fa5c84-c69a-49aa-e7d4-6f5450ed7591"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "In order to complete the configuration process, please execute the following:\n",
            "\n",
            "  echo \"export PATH=\\${PATH}:/root/edirect\" >> ${HOME}/.bashrc\n",
            "\n",
            "or manually edit the PATH variable assignment in your .bashrc file.\n",
            "\n",
            "Would you like to do that automatically now? [y/N]\n",
            "Holding off, then.\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Устонавливаем ещё программок"
      ],
      "metadata": {
        "id": "jQ6Shf2NL0wZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt update\n",
        "!apt install ncbi-blast+\n",
        "\n",
        "!wget https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
        "!tar -xzvf seqkit_linux_amd64.tar.gz\n",
        "!chmod a+x seqkit"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7EtGBtcKMIVL",
        "outputId": "261948de-1471-4182-9703-f05bf99717ad"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33m\r0% [Working]\u001b[0m\r            \rHit:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease\n",
            "\u001b[33m\r0% [Waiting for headers] [Waiting for headers] [Waiting for headers] [Waiting f\u001b[0m\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Waiting for headers] [Waiting for headers] [Wait\u001b[0m\r                                                                               \rIgn:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n",
            "\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Waiting for headers] [Waiting for headers] [Wait\u001b[0m\r                                                                               \rGet:3 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n",
            "\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Waiting for headers] [3 InRelease 14.2 kB/88.7 k\u001b[0m\r                                                                               \rHit:4 http://archive.ubuntu.com/ubuntu bionic InRelease\n",
            "\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Waiting for headers] [3 InRelease 14.2 kB/88.7 k\u001b[0m\r                                                                               \rHit:5 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease\n",
            "Ign:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n",
            "Hit:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release\n",
            "Hit:8 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n",
            "Get:9 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n",
            "Hit:10 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n",
            "Hit:11 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease\n",
            "Get:12 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]\n",
            "Hit:13 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n",
            "Fetched 252 kB in 2s (114 kB/s)\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "62 packages can be upgraded. Run 'apt list --upgradable' to see them.\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "ncbi-blast+ is already the newest version (2.6.0-1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 62 not upgraded.\n",
            "--2021-12-13 17:45:13--  https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.114.3\n",
            "Connecting to github.com (github.com)|140.82.114.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211213%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211213T174513Z&X-Amz-Expires=300&X-Amz-Signature=dbdb2274cca226b91b8f8da30bfcdc6f5bd8b5e03b3c6baa8d055dbd27cc2153&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-12-13 17:45:13--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211213%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211213T174513Z&X-Amz-Expires=300&X-Amz-Signature=dbdb2274cca226b91b8f8da30bfcdc6f5bd8b5e03b3c6baa8d055dbd27cc2153&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6650850 (6.3M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz.1’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   6.34M  --.-KB/s    in 0.1s    \n",
            "\n",
            "2021-12-13 17:45:13 (63.3 MB/s) - ‘seqkit_linux_amd64.tar.gz.1’ saved [6650850/6650850]\n",
            "\n",
            "seqkit\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем БД белков SwissProt"
      ],
      "metadata": {
        "id": "DvXwwE7zf0Qw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
        "!gzip -d uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6nOSPsdVf2Lo",
        "outputId": "b36a15e0-71b8-430b-8d63-6c9d1788d43d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-13 17:45:14--  https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
            "Resolving ftp.uniprot.org (ftp.uniprot.org)... 128.175.240.195\n",
            "Connecting to ftp.uniprot.org (ftp.uniprot.org)|128.175.240.195|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 90527596 (86M) [application/x-gzip]\n",
            "Saving to: ‘uniprot_sprot.fasta.gz’\n",
            "\n",
            "uniprot_sprot.fasta 100%[===================>]  86.33M   200MB/s    in 0.4s    \n",
            "\n",
            "2021-12-13 17:45:14 (200 MB/s) - ‘uniprot_sprot.fasta.gz’ saved [90527596/90527596]\n",
            "\n",
            "gzip: uniprot_sprot.fasta already exists; do you wish to overwrite (y or n)? ^C\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Теперь можно делать задание"
      ],
      "metadata": {
        "id": "PIBVoyeyf9UJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Cначало ищем гены"
      ],
      "metadata": {
        "id": "GXcCRnv6Nyum"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./gms2_linux_64/gms2.pl --seq scaffold.fasta --genome-type bacteria --fnn genes.fasta --faa proteins.fasta\n",
        "# мы получили файлы genes.fasta и proteins.fasta,\n",
        "# в которых соответственно гены и их белки.\n",
        "# это впринципе уже половина работы\n",
        "# осталось только понять что это за гены такие"
      ],
      "metadata": {
        "id": "DOPW6OIYHz0C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Потом пытаемся определить их назначение"
      ],
      "metadata": {
        "id": "QSwpykMdN2bu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# из родственной бактерии:\n",
        "!makeblastdb -dbtype prot -in T_oleivorans_MIL_1.proteins.fasta -out T_oleivorans_MIL_1.proteins\n",
        "!blastp -query proteins.fasta -db T_oleivorans_MIL_1.proteins -evalue 1e-10 -outfmt 6 > scaffolds.hits_from_MIL_1.txt\n",
        "\n",
        "# достаём только те гены, которых небыло в родственной бактерии\n",
        "!cut -f 1 scaffolds.hits_from_MIL_1.txt | sort -n | uniq > proteins.with_hits_from_MIL_1.txt\n",
        "!./seqkit grep --invert-match -f proteins.with_hits_from_MIL_1.txt proteins.fasta -o proteins.without_MIL_1.fasta\n",
        "\n",
        "# из большой базы данных:\n",
        "!makeblastdb -dbtype prot -in uniprot_sprot.fasta -out uniprot_sprot\n",
        "!blastp -query proteins.without_MIL_1.fasta -db uniprot_sprot -evalue 1e-10 -outfmt 6 > scaffolds.hits_from_SwissProt.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QfLgPcQVFRmf",
        "outputId": "d48b14e9-20e5-4c14-8b16-569fc8c09363"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/13/2021 14:39:10\n",
            "New DB name:   /content/T_oleivorans_MIL_1.proteins\n",
            "New DB title:  T_oleivorans_MIL_1.proteins.fasta\n",
            "Sequence type: Protein\n",
            "Deleted existing Protein BLAST database named /content/T_oleivorans_MIL_1.proteins\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 3666 sequences in 0.101555 seconds.\n",
            "[INFO]\u001b[0m 3334 patterns loaded from file\n",
            "\n",
            "\n",
            "Building a new DB, current time: 12/13/2021 14:40:13\n",
            "New DB name:   /content/uniprot_sprot\n",
            "New DB title:  uniprot_sprot.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 565928 sequences in 13.0161 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Теперь можно посчитать статистику"
      ],
      "metadata": {
        "id": "VdoQIFL0SFH3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "total_genes = !grep -c '>' genes.fasta\n",
        "total_genes = int(total_genes[0])\n",
        "\n",
        "rest_genes = !grep -c '>' proteins.without_MIL_1.fasta\n",
        "rest_genes = int(rest_genes[0])\n",
        "\n",
        "swiss_genes = !cut -f 1 scaffolds.hits_from_SwissProt.txt | sort | uniq | wc -l\n",
        "swiss_genes = int(swiss_genes[0])\n",
        "\n",
        "print(\"Cколько было предсказано генов всего?\", total_genes)\n",
        "print(\"Cколько из них удалось аннотировать с помощью сравнения с бактерией?\", total_genes-rest_genes)\n",
        "print(\"Сколько с помощью БД SwissProt?\", swiss_genes)\n",
        "print(\"Kакое кол-во белков остались без аннотации функции?\", rest_genes-swiss_genes)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nfNKGMRbSNfg",
        "outputId": "138c14e2-6a31-4af8-9d61-bba994fb1d7e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cколько было предсказано генов всего? 3616\n",
            "Cколько из них удалось аннотировать с помощью сравнения с бактерией? 3334\n",
            "Сколько с помощью БД SwissProt? 52\n",
            "Kакое кол-во белков остались без аннотации функции? 230\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Теперь осталось только залить на гитхаб"
      ],
      "metadata": {
        "id": "-9212_KzgIqw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!ssh-keygen"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5IpluCjcI_VG",
        "outputId": "1cc09599-3298-4f74-9b19-317cf8d5820c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Generating public/private rsa key pair.\n",
            "Enter file in which to save the key (/root/.ssh/id_rsa): \n",
            "Created directory '/root/.ssh'.\n",
            "Enter passphrase (empty for no passphrase): \n",
            "Enter same passphrase again: \n",
            "Your identification has been saved in /root/.ssh/id_rsa.\n",
            "Your public key has been saved in /root/.ssh/id_rsa.pub.\n",
            "The key fingerprint is:\n",
            "SHA256:99n9YMPtc7bHvhY3ryGIp7ltn/FHACsOMS8f0kXKsGg root@ab0bfbf39682\n",
            "The key's randomart image is:\n",
            "+---[RSA 2048]----+\n",
            "|        .  ..    |\n",
            "|       .o+ .o    |\n",
            "|      E .=o. o   |\n",
            "|     .  + = . .  |\n",
            "|        S*.o   . |\n",
            "|         ooo + =o|\n",
            "|        . o = BoB|\n",
            "|         =.  * BO|\n",
            "|        +o..o +BX|\n",
            "+----[SHA256]-----+\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat /root/.ssh/id_rsa.pub"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8RUfK1CNB-Cu",
        "outputId": "f95b388d-1215-4ce5-e922-83aca19e036d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCy61fM40N8jg1I6+7xWOm93VRrtYEpJbnvnS8J4/nb8+B8pD7LdPZGl2vkrA37h4+cmMEzs93vuJtUed9DAhJLtjceS2ibaj1AFNuAuE63gWlGyDOz1ZZP5XSxK0D9QRCfPsoKKQPP0cPs5f7/Zv79GdJBr8+RCGR0Jgh5kyP3QgRw6Ut74tGwZFPkuIu/Pl9XT1jsC5+lbBoEweI9+i7yh0SSvbIuntiUYZPCg5+Uu0R+94l0P2Zs/k9mnzFXlNIAkfb3znOOw8KH+zXBhtCnfFcy88WQ7J1orJkxqVLAjyh3cBU4b4xaXi2y7SSviiE5uUqkLlG7iPApgUgtdBYD root@ab0bfbf39682\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts\n",
        "!git clone git@github.com:Cortan122/hse21_hw2.git"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SUY0EcP1guMV",
        "outputId": "d89dd404-2a96-41d0-d64c-b73b8d21c93d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# github.com:22 SSH-2.0-babeld-fb957b4d\n",
            "Cloning into 'hse21_hw2'...\n",
            "Warning: Permanently added the RSA host key for IP address '140.82.114.3' to the list of known hosts.\n",
            "remote: Enumerating objects: 3, done.\u001b[K\n",
            "remote: Counting objects: 100% (3/3), done.\u001b[K\n",
            "remote: Compressing objects: 100% (2/2), done.\u001b[K\n",
            "remote: Total 3 (delta 0), reused 3 (delta 0), pack-reused 0\u001b[K\n",
            "Receiving objects: 100% (3/3), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir hse21_hw2/data\n",
        "!cp gms2.lst proteins.fasta scaffolds.hits_from_MIL_1.txt scaffolds.hits_from_SwissProt.txt hse21_hw2/data"
      ],
      "metadata": {
        "id": "ohkE64SQg5Ou"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!git config --global user.email \"knborisov@edu.hse.ru\"\n",
        "!git config --global user.name \"Костя Борисов (колаб)\""
      ],
      "metadata": {
        "id": "zRvBL60dhSkt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!cd hse21_hw2 && git add data && git commit -m \"данные из колаба\" && git push"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "JPmTNaf-hWm_",
        "outputId": "717858ba-7663-43ae-fea7-9623297765b5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[master 5199111] данные из колаба\n",
            " 4 files changed, 40359 insertions(+)\n",
            " create mode 100644 data/gms2.lst\n",
            " create mode 100644 data/proteins.fasta\n",
            " create mode 100644 data/scaffolds.hits_from_MIL_1.txt\n",
            " create mode 100644 data/scaffolds.hits_from_SwissProt.txt\n",
            "Warning: Permanently added the RSA host key for IP address '140.82.112.4' to the list of known hosts.\n",
            "Counting objects: 7, done.\n",
            "Delta compression using up to 2 threads.\n",
            "Compressing objects: 100% (7/7), done.\n",
            "Writing objects: 100% (7/7), 1.05 MiB | 5.88 MiB/s, done.\n",
            "Total 7 (delta 0), reused 0 (delta 0)\n",
            "To github.com:Cortan122/hse21_hw2.git\n",
            "   4bcb591..5199111  master -> master\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Вторая часть задания (про биопитон)"
      ],
      "metadata": {
        "id": "7-ZY36GXkJc1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ryrBgJMFhepr",
        "outputId": "1e5f332b-c2ce-49d2-897b-fc0490344d60"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 28.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython) (1.19.5)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.79\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
        "from datetime import datetime as dt\n",
        "import pandas as pd"
      ],
      "metadata": {
        "collapsed": true,
        "id": "znDLMRd-kkRm"
      },
      "execution_count": 101,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Подтягиваем геном"
      ],
      "metadata": {
        "id": "oHbqE8XKtlyC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# считываем все скаффолды в один большой словарь\n",
        "scaffolds = dict()\n",
        "for record in SeqIO.parse(\"scaffold.fasta\", \"fasta\"):\n",
        "  # проставляем всякие аннотации, которые от нас хочет genebank\n",
        "  # чтобы было красиво ✨\n",
        "  record.annotations['molecule_type'] = 'DNA'\n",
        "  record.annotations['organism'] = 'Thalassolituus oleivorans Barents'\n",
        "  record.annotations['source'] = 'Thalassolituus oleivorans Barents'\n",
        "  record.annotations['date'] = dt.now().strftime(\"%d-%b-%Y\").upper()\n",
        "  # record.annotations['topology'] = 'circular'\n",
        "  record.annotations['data_file_division'] = 'BCT'\n",
        "  \n",
        "  scaffolds[record.id] = record"
      ],
      "metadata": {
        "id": "F8HhC3XCjKXJ"
      },
      "execution_count": 103,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# файл gms2.lst очень сложно парсить\n",
        "# но нам он оказываестя не нужен (ура!)\n",
        "# вся информация дублируется в proteins.fasta\n",
        "genes = dict()\n",
        "for gene in SeqIO.parse(\"proteins.fasta\", \"fasta\"):\n",
        "  desc = gene.description.split(' ')\n",
        "  scaffold = desc[1]\n",
        "  start, end = int(desc[2]), int(desc[3])\n",
        "  strand = 1 if desc[4] == '+' else -1\n",
        "  \n",
        "  feat = SeqFeature(FeatureLocation(start, end, strand=strand), type=\"CDS\")\n",
        "  feat.qualifiers['locus_tag'] = [desc[0]]\n",
        "  feat.qualifiers['translation'] = [gene.seq]\n",
        "  scaffolds[scaffold].features.append(feat)\n",
        "  \n",
        "  genes[desc[0]] = feat"
      ],
      "metadata": {
        "id": "WbQKRYuljNMs"
      },
      "execution_count": 104,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Добавляем функции белков"
      ],
      "metadata": {
        "id": "oKW1zKcCttA5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# собираем функции всех генов родственной бактерии в ещё один словарь\n",
        "mil1genes = dict()\n",
        "for feat in SeqIO.read(\"T_oleivorans_MIL_1.gbk\", \"genbank\").features:\n",
        "  if 'protein_id' not in feat.qualifiers:\n",
        "    continue\n",
        "  if 'product' not in feat.qualifiers:\n",
        "    continue\n",
        "  \n",
        "  mil1genes[feat.qualifiers['protein_id'][0]] = feat.qualifiers['product'][0]"
      ],
      "metadata": {
        "id": "RemT_vdlYO-7"
      },
      "execution_count": 115,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# файл scaffolds.hits_from_MIL_1.txt это некий 6ой формат бласта\n",
        "# https://www.metagenomics.wiki/tools/blast/blastn-output-format-6\n",
        "# читаем его pandas-ом (🐼) потому что он табличка tsv\n",
        "names = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore']\n",
        "mil1hits = pd.read_csv('scaffolds.hits_from_MIL_1.txt', sep='\\t', header=None, names=names)\n",
        "mil1hits"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "collapsed": true,
        "id": "ZobO5urEphzi",
        "outputId": "5b0321d1-c0d6-4b61-b1f7-5d8c8377d2bb"
      },
      "execution_count": 152,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>qseqid</th>\n",
              "      <th>sseqid</th>\n",
              "      <th>pident</th>\n",
              "      <th>length</th>\n",
              "      <th>mismatch</th>\n",
              "      <th>gapopen</th>\n",
              "      <th>qstart</th>\n",
              "      <th>qend</th>\n",
              "      <th>sstart</th>\n",
              "      <th>send</th>\n",
              "      <th>evalue</th>\n",
              "      <th>bitscore</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71653.1_1204</td>\n",
              "      <td>98.990</td>\n",
              "      <td>99</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>8.080000e-68</td>\n",
              "      <td>196.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU72283.1_1834</td>\n",
              "      <td>97.980</td>\n",
              "      <td>99</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>2.290000e-67</td>\n",
              "      <td>194.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71933.1_1484</td>\n",
              "      <td>97.980</td>\n",
              "      <td>99</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>4.150000e-66</td>\n",
              "      <td>191.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71592.1_1143</td>\n",
              "      <td>97.980</td>\n",
              "      <td>99</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>1</td>\n",
              "      <td>99</td>\n",
              "      <td>4.150000e-66</td>\n",
              "      <td>191.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71865.1_1416</td>\n",
              "      <td>96.774</td>\n",
              "      <td>62</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>62</td>\n",
              "      <td>1</td>\n",
              "      <td>62</td>\n",
              "      <td>2.150000e-42</td>\n",
              "      <td>130.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10543</th>\n",
              "      <td>3614</td>\n",
              "      <td>lcl|HF680312.1_prot_1833</td>\n",
              "      <td>100.000</td>\n",
              "      <td>52</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>52</td>\n",
              "      <td>95</td>\n",
              "      <td>146</td>\n",
              "      <td>5.410000e-34</td>\n",
              "      <td>109.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10544</th>\n",
              "      <td>3614</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71654.1_1205</td>\n",
              "      <td>100.000</td>\n",
              "      <td>52</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>52</td>\n",
              "      <td>95</td>\n",
              "      <td>146</td>\n",
              "      <td>2.250000e-33</td>\n",
              "      <td>110.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10545</th>\n",
              "      <td>3614</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71593.1_1144</td>\n",
              "      <td>100.000</td>\n",
              "      <td>52</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>52</td>\n",
              "      <td>95</td>\n",
              "      <td>146</td>\n",
              "      <td>2.510000e-33</td>\n",
              "      <td>110.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10546</th>\n",
              "      <td>3614</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71934.1_1485</td>\n",
              "      <td>100.000</td>\n",
              "      <td>52</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>52</td>\n",
              "      <td>95</td>\n",
              "      <td>146</td>\n",
              "      <td>4.490000e-33</td>\n",
              "      <td>110.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10547</th>\n",
              "      <td>3614</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71223.1_774</td>\n",
              "      <td>100.000</td>\n",
              "      <td>30</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>30</td>\n",
              "      <td>95</td>\n",
              "      <td>124</td>\n",
              "      <td>6.380000e-16</td>\n",
              "      <td>63.5</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>10548 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       qseqid                               sseqid  ...        evalue  bitscore\n",
              "0           1  lcl|HF680312.1_prot_CCU71653.1_1204  ...  8.080000e-68     196.0\n",
              "1           1  lcl|HF680312.1_prot_CCU72283.1_1834  ...  2.290000e-67     194.0\n",
              "2           1  lcl|HF680312.1_prot_CCU71933.1_1484  ...  4.150000e-66     191.0\n",
              "3           1  lcl|HF680312.1_prot_CCU71592.1_1143  ...  4.150000e-66     191.0\n",
              "4           1  lcl|HF680312.1_prot_CCU71865.1_1416  ...  2.150000e-42     130.0\n",
              "...       ...                                  ...  ...           ...       ...\n",
              "10543    3614             lcl|HF680312.1_prot_1833  ...  5.410000e-34     109.0\n",
              "10544    3614  lcl|HF680312.1_prot_CCU71654.1_1205  ...  2.250000e-33     110.0\n",
              "10545    3614  lcl|HF680312.1_prot_CCU71593.1_1144  ...  2.510000e-33     110.0\n",
              "10546    3614  lcl|HF680312.1_prot_CCU71934.1_1485  ...  4.490000e-33     110.0\n",
              "10547    3614   lcl|HF680312.1_prot_CCU71223.1_774  ...  6.380000e-16      63.5\n",
              "\n",
              "[10548 rows x 12 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 152
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# в этой табличке по нескольку кандидатов для каждого гена\n",
        "# надо выбрать лучшее совпадаение для каждого\n",
        "# а хорошесть совпадения это bitscore\n",
        "\n",
        "# также надо выкинуть стоки в которых нет \"CCU\"\n",
        "# потому что у нас нету таких в базе данных значаний генов\n",
        "# наеврно эту чтения попали на межгенное пространство 🤔\n",
        "hits = mil1hits[mil1hits['sseqid'].str.contains(\"CCU\")].sort_values('bitscore', ascending=False).drop_duplicates('qseqid')\n",
        "hits"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "collapsed": true,
        "id": "giiH5b9RvUty",
        "outputId": "56e8d56c-6f65-483b-cffb-5e56ef65f004"
      },
      "execution_count": 144,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>qseqid</th>\n",
              "      <th>sseqid</th>\n",
              "      <th>pident</th>\n",
              "      <th>length</th>\n",
              "      <th>mismatch</th>\n",
              "      <th>gapopen</th>\n",
              "      <th>qstart</th>\n",
              "      <th>qend</th>\n",
              "      <th>sstart</th>\n",
              "      <th>send</th>\n",
              "      <th>evalue</th>\n",
              "      <th>bitscore</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2831</th>\n",
              "      <td>1023</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU73896.1_3447</td>\n",
              "      <td>98.979</td>\n",
              "      <td>2253</td>\n",
              "      <td>23</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>2253</td>\n",
              "      <td>1</td>\n",
              "      <td>2253</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>4296.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4238</th>\n",
              "      <td>1485</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU70724.1_275</td>\n",
              "      <td>97.870</td>\n",
              "      <td>2113</td>\n",
              "      <td>43</td>\n",
              "      <td>1</td>\n",
              "      <td>7</td>\n",
              "      <td>2119</td>\n",
              "      <td>1</td>\n",
              "      <td>2111</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>4158.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9135</th>\n",
              "      <td>3215</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71900.1_1451</td>\n",
              "      <td>99.878</td>\n",
              "      <td>1640</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1640</td>\n",
              "      <td>1</td>\n",
              "      <td>1640</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>3406.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2609</th>\n",
              "      <td>947</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU73861.1_3412</td>\n",
              "      <td>99.214</td>\n",
              "      <td>1654</td>\n",
              "      <td>13</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1654</td>\n",
              "      <td>1</td>\n",
              "      <td>1654</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>3372.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4001</th>\n",
              "      <td>1452</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU70690.1_241</td>\n",
              "      <td>99.392</td>\n",
              "      <td>1645</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1645</td>\n",
              "      <td>1</td>\n",
              "      <td>1645</td>\n",
              "      <td>0.000000e+00</td>\n",
              "      <td>3338.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9904</th>\n",
              "      <td>3484</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU71629.1_1180</td>\n",
              "      <td>61.224</td>\n",
              "      <td>49</td>\n",
              "      <td>8</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>39</td>\n",
              "      <td>139</td>\n",
              "      <td>187</td>\n",
              "      <td>1.280000e-11</td>\n",
              "      <td>54.7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10388</th>\n",
              "      <td>3591</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU73193.1_2744</td>\n",
              "      <td>29.592</td>\n",
              "      <td>98</td>\n",
              "      <td>63</td>\n",
              "      <td>2</td>\n",
              "      <td>16</td>\n",
              "      <td>112</td>\n",
              "      <td>4</td>\n",
              "      <td>96</td>\n",
              "      <td>1.250000e-11</td>\n",
              "      <td>53.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8044</th>\n",
              "      <td>2814</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU73711.1_3262</td>\n",
              "      <td>29.592</td>\n",
              "      <td>98</td>\n",
              "      <td>63</td>\n",
              "      <td>2</td>\n",
              "      <td>16</td>\n",
              "      <td>112</td>\n",
              "      <td>4</td>\n",
              "      <td>96</td>\n",
              "      <td>1.250000e-11</td>\n",
              "      <td>53.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4800</th>\n",
              "      <td>1680</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU70460.1_11</td>\n",
              "      <td>29.592</td>\n",
              "      <td>98</td>\n",
              "      <td>63</td>\n",
              "      <td>2</td>\n",
              "      <td>16</td>\n",
              "      <td>112</td>\n",
              "      <td>4</td>\n",
              "      <td>96</td>\n",
              "      <td>1.250000e-11</td>\n",
              "      <td>53.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5991</th>\n",
              "      <td>2116</td>\n",
              "      <td>lcl|HF680312.1_prot_CCU73039.1_2590</td>\n",
              "      <td>100.000</td>\n",
              "      <td>28</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>28</td>\n",
              "      <td>1</td>\n",
              "      <td>28</td>\n",
              "      <td>5.730000e-12</td>\n",
              "      <td>50.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3334 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       qseqid                               sseqid  ...        evalue  bitscore\n",
              "2831     1023  lcl|HF680312.1_prot_CCU73896.1_3447  ...  0.000000e+00    4296.0\n",
              "4238     1485   lcl|HF680312.1_prot_CCU70724.1_275  ...  0.000000e+00    4158.0\n",
              "9135     3215  lcl|HF680312.1_prot_CCU71900.1_1451  ...  0.000000e+00    3406.0\n",
              "2609      947  lcl|HF680312.1_prot_CCU73861.1_3412  ...  0.000000e+00    3372.0\n",
              "4001     1452   lcl|HF680312.1_prot_CCU70690.1_241  ...  0.000000e+00    3338.0\n",
              "...       ...                                  ...  ...           ...       ...\n",
              "9904     3484  lcl|HF680312.1_prot_CCU71629.1_1180  ...  1.280000e-11      54.7\n",
              "10388    3591  lcl|HF680312.1_prot_CCU73193.1_2744  ...  1.250000e-11      53.5\n",
              "8044     2814  lcl|HF680312.1_prot_CCU73711.1_3262  ...  1.250000e-11      53.5\n",
              "4800     1680    lcl|HF680312.1_prot_CCU70460.1_11  ...  1.250000e-11      53.5\n",
              "5991     2116  lcl|HF680312.1_prot_CCU73039.1_2590  ...  5.730000e-12      50.1\n",
              "\n",
              "[3334 rows x 12 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 144
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for i, hit in hits.iterrows():\n",
        "  gene = genes[str(hit['qseqid'])]\n",
        "  match = hit['sseqid'].split('_')[2]\n",
        "  gene.qualifiers['product'] = [mil1genes[match]]"
      ],
      "metadata": {
        "id": "OZk4fyWFzUEy"
      },
      "execution_count": 147,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Теперь тоже самое для SwissProt"
      ],
      "metadata": {
        "id": "Pov7-nfp6z2x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -nc https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz\n",
        "!gzip -d uniprot_sprot.dat.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a-ppLQ8n7Upp",
        "outputId": "3f006d65-862c-4651-dd58-e33f156f0ebe"
      },
      "execution_count": 156,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File ‘uniprot_sprot.dat.gz’ already there; not retrieving.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# файл uniprot_sprot.dat очень большой (3.5GB)\n",
        "# питон с таким файлом несправится\n",
        "# нужно достать из него самое нужное grep-ом\n",
        "!grep '^ID\\|^DE   RecName: Full=' uniprot_sprot.dat > SwissProt_names.txt"
      ],
      "metadata": {
        "id": "qwPTBDDt9-dP"
      },
      "execution_count": 173,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "previd = None\n",
        "swissgenes = dict()\n",
        "for line in open('SwissProt_names.txt'):\n",
        "  if line.startswith('ID'):\n",
        "    previd = line.split()[1]\n",
        "  if line.startswith('DE'):\n",
        "    swissgenes[previd] = line.split('=')[1][:-2]"
      ],
      "metadata": {
        "id": "bFYd_-Al-OYs"
      },
      "execution_count": 181,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "swisshits_raw = pd.read_csv('scaffolds.hits_from_SwissProt.txt', sep='\\t', header=None, names=names)\n",
        "swisshits = swisshits_raw.sort_values('bitscore', ascending=False).drop_duplicates('qseqid')\n",
        "\n",
        "for i, hit in swisshits.iterrows():\n",
        "  gene = genes[str(hit['qseqid'])]\n",
        "  match = hit['sseqid'].split('|')[-1]\n",
        "  gene.qualifiers['product'] = [swissgenes[match]]"
      ],
      "metadata": {
        "id": "_OQY-l1P6965"
      },
      "execution_count": 187,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Genbank готов, заливаем его на гитхаб"
      ],
      "metadata": {
        "id": "pwI6yffkBKpR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "SeqIO.write(scaffolds.values(), \"GENOME.gbk\", \"genbank\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "t2QuHFbElPpK",
        "outputId": "f1cc1f17-9410-4c03-d7b5-3d8c1b1c0510"
      },
      "execution_count": 188,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "73"
            ]
          },
          "metadata": {},
          "execution_count": 188
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cp GENOME.gbk hse21_hw2/data\n",
        "!cd hse21_hw2 && git add data && git commit -m \"геном из колаба\" && git push"
      ],
      "metadata": {
        "id": "3plA4_h00To6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4a36534d-e973-4ab0-b337-41b62ed300e0"
      },
      "execution_count": 191,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[master c1725bf] геном из колаба\n",
            " 1 file changed, 99621 insertions(+)\n",
            " create mode 100644 data/GENOME.gbk\n",
            "Warning: Permanently added the RSA host key for IP address '140.82.114.4' to the list of known hosts.\n",
            "Counting objects: 4, done.\n",
            "Delta compression using up to 2 threads.\n",
            "Compressing objects: 100% (4/4), done.\n",
            "Writing objects: 100% (4/4), 2.41 MiB | 1.78 MiB/s, done.\n",
            "Total 4 (delta 1), reused 0 (delta 0)\n",
            "remote: Resolving deltas: 100% (1/1), completed with 1 local object.\u001b[K\n",
            "To github.com:Cortan122/hse21_hw2.git\n",
            "   5199111..c1725bf  master -> master\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "3JYtRXjiB704"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}